### ISSUE TITLE
Migrate/Visit items in a processor

### ISSUE THREAD

**Issue by frairon** [lovoo/goka#329]
I'm proposing a new feature, that allows to iterate all items of a processor table in order to modify them somehow.

The idea came from the following problem: Let's say we start with a processor accumulating messages
```go
type Messages struct{
  messages []string
}
```

```go
func consume(ctx goka.Context, msg interface{}){
  var msgs *Messages
  if val := ctx.Value(); val != nil{
      msgs = val.(*Messages)
  }else{
      msgs =  new(Messages)
  }
  // append messages
  msgs.messages = append(msgs.messages, msg.(string))
 
  ctx.SetValue(msgs)
}```


Then our application evolves, and we realize that we actually wanted to count the messages instead.

So let's extend our data model:
```go
type Messages struct{
  messages []string
  count int
}
```

Now we have two options:

### Option 1: migrate forever
we set the count-field and remove the messages every time we touch the value. So we'll do an online-migration. E.g. like this:

our migrating handler could look like this
```go
func consume(ctx goka.Context, msg interface{}){
  // same as before
  var msgs *Messages
  if val := ctx.Value(); val != nil{
      msgs = val.(*Messages)
  }else{
      msgs =  new(Messages)
  }
  
  // migrate from previous messages to the count
  if len(msgs.messages) != 0{
     msgs.count = len(msgs.messages)
     msgs.messages = nil
  }
  
  // increment the counter instead of adding the message:
  msgs.count++

  // store back
  ctx.SetValue(msgs)
}
```

Pros:
* easy to implement at first

Cons:
* all "downstream"-components like views or other processors have to consider the two datastrcutures, so we have to stay backwards-compatible forever
* migration is never done, we'll always have this "legacy"-field and cannot get rid of it

### Option 2: migration once
Instead of migrating forever, we'll implement a "migrate-function", that only migrates:

```go
func migrate(ctx goka.Context, _ interface{}){
   if val := ctx.Value(); val != nil{
      msgs := val.(*Messages)
      // migrate from previous messages to the count
      if len(msgs.messages) != 0{
         msgs.count = len(msgs.messages)
         msgs.messages = nil
      }
      ctx.SetValue(msgs)
    }
}
```
Then we'll create a new kafka-topic and make the processor run that callback for messages in this topic, like this
```go
proc := goka.NewProcessor(...
   goka.DefineGroup(...
      goka.Input("message-migration-topic", new(codec.String), migrate),
    ),
)
```
But how do we get to iterate over all messages?
For this, we have to create a view and once it's recovered, iterate over it and send all keys to the topic to migrate. I'll spare you the code example on this one :) 

Pro:
* migration is done once

Cons:
* extra topic necessary
* lots to implement
* create an extra view just to iterate once, then throw it away again.

Of course this example is not very realistic and there might be other solutions as well, but we've faced situations like these a couple of times and were forced to use option2.
So here's a new proposal:

### Proposal
Migrate within the processor. We're configuring the processor (via group graph) to support some migration-logic, like this:

```go
proc := goka.NewProcessor(...
   goka.DefineGroup(...
      goka.Visitor("message-count-migration", func(ctx goka.Context, msg interface{}){
            if val := ctx.Value(); val != nil{
		 msgs := val.(*Messages)
		// migrate from previous messages to the count
		if len(msgs.messages) != 0{
		 msgs.count = len(msgs.messages)
		 msgs.messages = nil
		}
		ctx.SetValue(msgs)
	   }
	}),
    ),
)
``` 

The passed callback is has the same signature as a normal event-callback, but it's input event will not be a kafka message but a user-provided value (or nil). The elements of the context (Emit, Join, Lookup, Value/SetValue) are available as usual.

Furthermore the processor supports a new function: `Visit` which allows to visit each value once, passing a single value, if needed
```go
proc.Visit(context.Background(), "message-count-migration", nil)
```
Internally, it will trigger all partition processors to iterate their tables and call the migration function for each value.

The proof-of-concept implementation is in the branch [visitor-tool-poc](https://github.com/lovoo/goka/tree/visitor-tool-poc)
An example use case is in [10-visit](https://github.com/lovoo/goka/blob/visitor-tool-poc/examples/10-visit/main.go)

There are still some unclear things here to figure out, especially regarding shutdown/rebalance and triggering the visit, but the general PoC is working.
What do you think?

**Comment by db7** [comment]
This will only change the leveldb data or will visit also update the table topic in kafka? What about views on this table topic? They will have to implement that as well? I have the feeling that as an external observer of the topic, one may see the type of a key jumping back and forth.

**Comment by frairon** [comment]
@db7 The visit function is a normal callback with the same functionality as a normal callback, but instead of receiving the message from kafka, it's generated internally for each key in the table. That way, a call to `ctx.SetValue()` updates leveldb and also sends the changes to kafka and to all views. It's also fine to emit new messages with `ctx.Emit`, or do `ctx.Join/Lookup`, because the function is called in a running instance.
If the processor is scaled across multiple instances, the visit needs to be initiated on all instances, since the iteration on each partition separately.

At first, we'll mostly use this to migrate state, like removing some legacy data.
In future however there are more use cases, e.g. we could use it to do data-cleanup by iterating over the processor state and remove old items (calling  `ctx.Delete()` is also fine). In this case, a ticker could trigger the "cleanup" on the processor table on a nightly basis.

Or another idea: if the table is small enough (or there are enough partitions), maybe one could even use it to schedule an action on items of a table, like regularily emitting a window and resetting the state. But since the iteration might take a while this could lead to quite some inaccurate results.

Summary: all changes are send downstream as if the callback was called from a normal input-message, and it is semantically equivalent of having a separate view and sending a message for each key to the upstream processor - just less implementation overhead.

**Comment by ubunatic** [comment]
How do you ensure the visitor does not cause data races? For instance:
```
// 1. the processor reads a non-migrated value           ...
msgs := val.(*Messages)                                  ...
...                                                      // 2. the visitor reads and sets a new migrated value
...                                                      ctx.SetValue(msgs)
// 3. the processor updates a non-migrated value         ...
ctx.SetValue(msgs)                                       ...
```

**Comment by frairon** [comment]
It cannot have data races for the same reason that goka does not have data races when normally processing input events: only partitions are processed in parallel, within one partition, all events are processed sequentially and they don't share a storage. 
The visit function will be triggered by an "artificial event" that is added to the sequence like normal input events from Kafka.

**Comment by frairon** [comment]
Feature implemented in #341 and #345, closing the issue
